crates/binjs_io/src/entropy/predict.rs

```diff
@@ -336,3 +344,192 @@ impl<NodeValue> PathPredict<NodeValue, SymbolInfo> where NodeValue: Eq + Hash +
     }
 }
 
+
+
+
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Serialize, Deserialize)]
+enum WindowPrediction {
+    BackReference(BackReference),
+    DictionaryIndex(DictionaryIndex),
+}
+
+/// A prediction mechanism based on a sliding window.
+///
+/// Whenever encoding/decoding a value, if this value is one of the `width` latest
+/// seen values, we use a backreference, otherwise, we use an index into a global
+/// list of values. This strategy should work well when backreferences have a high
+/// probability of happening.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct WindowPredict<NodeValue, Statistics> where NodeValue: Clone + Eq + Hash {
+    /// The window width.
+    width: usize,
+
+    /// The `width` values seen most recently, indexed by `BackReference`. Note that
+    /// this vector starts empty and may contain fewer than `width` values. A value
+    /// may never appear twice in this vector.
+    latest_values: Vec<NodeValue>,
+
+    /// A dictionary of all values encountered so far.
+    /// Whenever we encounter a value that had previously not been encountered,
+    /// we append it at the end. Values never move.
+    value_by_dictionary_index: Vec<NodeValue>,
+
+    /// A mapping from NodeValue to the `DictionaryIndex` used to represent them
+    /// in `value_by_dictionary_index`. Mapping for a given value never changes.
+    dictionary_index_by_value: HashMap<NodeValue, DictionaryIndex>,
+
+    /// Actual statistics on values.
+    info: ContextInformation<WindowPrediction, Statistics>,
+}
+impl<NodeValue, Statistics> WindowPredict<NodeValue, Statistics> where NodeValue: Clone + Eq + Hash {
+    pub fn new(width: usize) -> Self {
+        WindowPredict {
+            width,
+            value_by_dictionary_index: Vec::with_capacity(1024), // FIXME: Magic number
+            dictionary_index_by_value: HashMap::with_capacity(1024),
+            latest_values: Vec::with_capacity(width),
+            info: ContextInformation::new(),
+        }
+    }
+
+    /// Update current window by moving the value at index `index` to the latest-seen
+    /// position (index 0).
+    ///
+    /// Fails if the index is not in [0, self.latest_values.len()[
```

Comment:
this has failure branch instead of assertion because the `index` is supposed to be raw value read from external file and can be outside the range, right?